packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -0,0 +1,208 @@
+                    )
+                )
+            );
+
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signature.length > 0) {
+                require(txHash.verifySignature(withdrawal.owner, auxData.signature), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[withdrawal.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[withdrawal.owner][txHash] = false;
+            }
+        } else if (withdrawal.withdrawalType == 2 || withdrawal.withdrawalType == 3) {
+            // Forced withdrawals cannot make use of certain features because the
+            // necessary data is not authorized by the account owner.
+            require(withdrawal.owner == withdrawal.to, "INVALID_WITHDRAWAL_ADDRESS");
+            require(withdrawal.fee == 0, "FEE_NOT_ZERO");
+            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+
+            ExchangeData.ForcedWithdrawal storage forcedWithdrawal = S.pendingForcedWithdrawals[withdrawal.accountID][withdrawal.tokenID];
+            if (forcedWithdrawal.timestamp == 0) {
+                // Allow the operator to submit fill withdrawals without authorization
+                // - when in shutdown mode
+                // - to withdraw protocol fees
+                require(withdrawal.accountID == 0 || S.isShutdown(), "FULL_WITHDRAWAL_UNAUTHORIZED");
+            } else {
+                // Type == 2: valid onchain withdrawal started by the owner
+                // Type == 3: invalid onchain withdrawal started by someone else
+                bool authorized = (withdrawal.owner == forcedWithdrawal.owner);
+                require((withdrawal.withdrawalType == 2) == authorized, "INVALID_WITHDRAW_TYPE");
+                if (!authorized) {
+                    require(withdrawal.amount == 0, "UNAUTHORIZED_WITHDRAWAL");
+                }
+
+                // Get the fee
+                feeETH = forcedWithdrawal.fee;
+
+                // Reset the approval so it can't be used again
+                S.pendingForcedWithdrawals[withdrawal.accountID][withdrawal.tokenID] = ExchangeData.ForcedWithdrawal(address(0), 0, 0);
+
+                // Open up a slot
+                S.numPendingForcedTransactions--;
+
+                emit ForcedWithdrawalConsumed(withdrawal.accountID, withdrawal.tokenID, withdrawal.amount);
+            }
+        } else {
+            revert("INVALID_WITHDRAWAL_TYPE");
+        }
+
+        // Validate gas provided
+        require(auxData.gasLimit >= withdrawal.minGas, "INVALID_GAS_AMOUNT");
+        // Validate the auxixliary withdrawal data
+        if (withdrawal.dataHash == bytes32(0)) {
+            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+        } else {
+            require((uint(keccak256(auxData.auxiliaryData)) >> 3) == uint(withdrawal.dataHash), "INVALID_WITHDRAWAL_AUX_DATA");
```

Comment:
The SNARK field can store 253 bits (the max value is in the order of 254 bits), so we have to go from 256 bit hashes to 253 bit hashes so that we can store a hash value inside a single field element.